### PR TITLE
Unravel CH4 

### PR DESCRIPTION
**Pull request by toniseibold**
Closes https://github.com/PyPSA/pypsa-ariadne/issues/45

Checkout https://github.com/PyPSA/pypsa-eur/pull/1159 for necessary pypsa-eur changes
Also needs changes from https://github.com/PyPSA/pypsa-eur/pull/1162

## Method
In the rule 'modify_prenetwork` the following things are happening:
- new buses for `consumption gas` and `renewable gas` are added to each node
- the gas stores are copied to allow storing renewable gas as well
- links from `renewable gas` and `gas` to `consumption gas` are added for each node
- the links consuming `gas` are connected to the `consumption gas` bus
- for the gas pipelines, two cases have to be covered
   - `config["sector"]["H2_retrofit"]: true`
   - only the gas pipelines from the current `planning horizon` exist ('add_brownfield`). Their installed capacity is already reduced by the installed capacity of retrofitted hydrogen pipelines.
   - `config["sector"]["H2_retrofit"]: false`
   - renewable gas pipelines are deleted and added again. This is necessary since the installed capacity of renewable gas pipelines is in some cases below the threshold capacity in `add_brownfield`
- the new gas pipelines are added for renewables by copying them only for the current planning_horizon
- the installed capacities of new (renewable) gas pipelines from the previous planning_horizon are matched

In the rule `additional functionality` the following boundary conditions are added:
- The stored fossil and renewable gas can not exceed the installed capacity of the fossil gas stores
- for the gas pipelines, two cases have to be covered
   - `config["sector"]["H2_retrofit"]: true`
   - the gas pipelines only from the first planning horizon exist and are extendable. Therefore, a simple boundary condition limiting the transport in the renewable and fossil being smaller than the installed capacity is needed
   - `config["sector"]["H2_retrofit"]: false`
   - the current planning horizon infrastructure is extendable and therefore the flow in renewable and fossil gas pipeline must be lower equal than the installed capacity of the fossil pipeline. This also covers the edge case if the system would want to enhance only renewable gas pipelines which would later be deleted in `modify_prenetwork`
   - for the already existing gas infrastructure the flow from renewable and fossil gas should not exceed the installed capacity
- the power flow in the new (renewable) gas pipelines from the previous planning_horizon is limited to the installed capacity

### CODE REVIEW

**Comment by toniseibold** [comment]
Still having trouble with the boundary condition if `H2_retrofit: true` for existing gas pipelines. But other than that done.
@lindnemi: can you take care of the exporter?

File: workflow/scripts/modify_prenetwork.py [hunk 1]
```diff
@@ -336,6 +336,173 @@ def unravel_oilbus(n):
...
+    XX1 0 gas:              fossil gas bus
+    XX1 0 gas renewable:    renewable gas bus
+
+    Renewable gas from biomass or methanation processes feed into the renewable gas bus.
+    The pipeline links are copied as well to allow the transport/trade of fossil and renewable gas separately.
+    Additionally, stores are added to enable the storage of fossil and renewable gas.
+    The boundary conditions are added in the additional_functionality script.
+    """
+
+    consumption_carriers = ["urban central gas boiler",
```
[the remark this answers is not in the document]
Comment (lindnemi) [reply workflow/scripts/modify_prenetwork.py:352]: hab es hinzugefügt

File: workflow/scripts/additional_functionality.py [hunk 2]
```diff
@@ -411,13 +419,69 @@ def add_h2_derivate_limit(n, snapshots, investment_year, config):
...
+        # if H2 retrofit is not active, gas pipelines for each planning_horizon exist
+        # the current planning horizon is p_nom_extendable
+        if not fossil_gas_pipe[~fossil_gas_pipe.p_nom_extendable].empty:
+            lhs = n.model["Link-p"].loc[:, fossil_gas_pipe[~fossil_gas_pipe.p_nom_extendable].index] + n.model["Link-p"].loc[:, re_gas_pipe[~re_gas_pipe.p_nom_extendable].index]
+            rhs = n.links.p_nom.loc[fossil_gas_pipe[~fossil_gas_pipe.p_nom_extendable].index]
+            n.model.add_constraints(lhs <= rhs, name=f"non extendable fossil+renewable gas pipelines")
+
+        else:
+            lhs = n.model["Link-p"].loc[:, fossil_gas_pipe[fossil_gas_pipe.p_nom_extendable].index] + n.model["Link-p"].loc[:, re_gas_pipe[re_gas_pipe.p_nom_extendable].index]
+            rhs = n.model["Link-p_nom"].loc[fossil_gas_pipe[fossil_gas_pipe.p_nom_extendable].index]
```
Comment (lindnemi) [inline workflow/scripts/additional_functionality.py:450]: why is this p_nom instead of p_nom_opt?

**Comment by lindnemi** [comment]
I just added SMR and SMR CC to the consumption carriers and merged main/master here and on pypsa-eur. 

Now the solver reports `infeasible_or_unbounded` and afterwards goes into an optimization effort which just goes on and on:

image

After >10 minutes i killed it because it should not take longer than 1 minute:

image

**Pull request closed without merging**